# Исследование надежности заемщиков.

 Надежность заемщиков является основной причиной выдачи кредитов.В зависимости от семейного положения, возраста, количества или отсутствия дейтей меняется и срок возврата заемных средств. Далее расмотрим ,как изменяется зависимость от должности, возраста, места работы, количество детей и прочим факторам, и возврата крелита в срок. 

### Шаг 1. Обзор данных

In [2]:
import pandas as pd

In [4]:
data = pd.read_csv('C:/Users/Жора/Downloads/проекты/готовое/2й предобработка данных/data.csv')
data.info()
#display(data)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
display(data.head())

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

После запроса информации по таблице видно что в стобцах days_employed и total_income меньше значений 19351, тогда как в других по 21525.

In [6]:
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

### Шаг 2.1 Заполнение пропусков

Проверю сколько пропущенных значений в столбцах days_employed и total_income.

In [7]:
display(data.isna().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В столбце total_income присутствует 2174 пропущенных значения(Nan). В total_income находится информация 
о заработной плате человека, и так как значений не мало их проще заменить на медианное значение. так как среднее значение здесь
не подойдет, потому что заработная плата у многих разная.

In [8]:
median_total = data['total_income'].median()
print(median_total)
data['total_income']=data['total_income'].fillna(median_total)
display(data[data['days_employed'].isna()])
print(data['debt'].isna().sum())

145017.93753253992


children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

        total_income                           purpose  
12     145017.937533                   сыграть свадьбу  
26     145017.937533                       образование  
29     145017.937533  строительство жилой недвижимости  
41     145017.937533  сделка с подержанным автомобилем  
55     145017.937533                   сыграть свадьбу  
...              ...                               ...  
21489  145017.937533              сделка с автомобилем  
21495  145017.937533                           свадьба  
21497  145017.937533        строительство недвижимости  
21502  145017.937533  строительство жилой недвижимости  
21510  145017.937533           приобретение автомобиля  

[2174 rows x 12 columns]

0


Поиск пропусков в зависимости от пола, гендера, тип занятости

In [9]:
gender = data['gender'].isna().sum()
display(f'Пропущенных значений в графе по полу -{gender}')
dob_years = data['dob_years'].isna().sum()
display(f'Пропущенных значений в графе по возрасту -{dob_years}')
income_type=data['income_type'].isna().sum()
display(f'Пропущенных значений в графе по типу занятости -{income_type}')

'Пропущенных значений в графе по полу -0'

'Пропущенных значений в графе по возрасту -0'

'Пропущенных значений в графе по типу занятости -0'

In [10]:
data_mean=(data['days_employed'].min() + data['days_employed'].max())/2
data['days_employed']=data['days_employed'].fillna(data_mean)
display(data.isna().sum())

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Шаг 2.2 Проверка данных на аномалии и исправления.

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [12]:
display(data.value_counts())

children  days_employed   dob_years  education  education_id  family_status          family_status_id  gender  income_type  debt  total_income   purpose                               
0          191683.225287  47         среднее    1             женат / замужем        0                 F       сотрудник    0     145017.937533  ремонт жилью                              3
                          62         среднее    1             женат / замужем        0                 F       пенсионер    0     145017.937533  ремонт жилью                              3
                          47         среднее    1             женат / замужем        0                 F       сотрудник    0     145017.937533  строительство недвижимости                2
                          23         среднее    1             Не женат / не замужем  4                 F       сотрудник    0     145017.937533  сделка с подержанным автомобилем          2
                          35         среднее    1           

### Шаг 2.3. Изменение типов данных.

Из запрощенной информации по таблице видно что у двух столбцов занчения float, заменю их на int, так удобнее для чтения и работы.

In [13]:
data['total_income']= data['total_income'].astype('int32')
data['days_employed']=data['days_employed'].astype('int32')
data['children']=data['children'].astype('int64')
data['education_id']=data['education_id'].astype('int8')
data['family_status_id']=data['family_status_id'].astype('int8')
data['dob_years']=data['dob_years'].astype('int32')
data['debt']=data['debt'].astype('int8')
data.info()
data['children'].value_counts()
#display(data)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int32 
 2   dob_years         21525 non-null  int32 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int8  
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int8  
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int8  
 10  total_income      21525 non-null  int32 
 11  purpose           21525 non-null  object
dtypes: int32(3), int64(1), int8(3), object(5)
memory usage: 1.3+ MB


 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

### Шаг 2.4. Удаление дубликатов.

In [14]:
display(data['education'].duplicated().sum())
display(data)
data['children']=data['children'].replace(-1,1)
data['children']=data['children'].replace(20,2)
print(data['children'].value_counts())
data=data.drop_duplicates().reset_index(drop=True)
display(data)

21510

children  days_employed  dob_years education  education_id  \
0             1          -8437         42    высшее             0   
1             1          -4024         36   среднее             1   
2             0          -5623         33   Среднее             1   
3             3          -4124         32   среднее             1   
4             0         340266         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1          -4529         43   среднее             1   
21521         0         343937         67   среднее             1   
21522         1          -2113         38   среднее             1   
21523         3          -3112         38   среднее             1   
21524         2          -1984         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21520        224791             операции с жильем  
21521        155999          сделка с автомобилем  
21522         89672                  недвижимость  
21523        244093  на покупку своего автомобиля  
21524         82047         на покупку автомобиля  

[21525 rows x 12 columns]

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64


children  days_employed  dob_years education  education_id  \
0             1          -8437         42    высшее             0   
1             1          -4024         36   среднее             1   
2             0          -5623         33   Среднее             1   
3             3          -4124         32   среднее             1   
4             0         340266         53   среднее             1   
...         ...            ...        ...       ...           ...   
21466         1          -4529         43   среднее             1   
21467         0         343937         67   среднее             1   
21468         1          -2113         38   среднее             1   
21469         3          -3112         38   среднее             1   
21470         2          -1984         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21466  гражданский брак                 1      F   компаньон     0   
21467   женат / замужем                 0      F   пенсионер     0   
21468  гражданский брак                 1      M   сотрудник     1   
21469   женат / замужем                 0      M   сотрудник     1   
21470   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21466        224791             операции с жильем  
21467        155999          сделка с автомобилем  
21468         89672                  недвижимость  
21469        244093  на покупку своего автомобиля  
21470         82047         на покупку автомобиля  

[21471 rows x 12 columns]

Следующим шагом проверю таблицу на неявные повторы, как видно ниже в основном изменен регистр в графе об образовании и
в столбце при вызове уникальных занчений по возрасту нашелся человек с возратсом 0, это явно ошибка.

In [15]:
print(data['education'].value_counts())

print(data['family_status'].value_counts())

среднее                13705
высшее                  4710
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   273
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64
женат / замужем          12344
гражданский брак          4163
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64


In [16]:
data['education']=data['education'].str.lower()
data['family_status']=data['family_status'].str.lower()
display(data['education'].value_counts())
display(data['family_status'].value_counts())
data=data.drop_duplicates().reset_index(drop=True)
display(data)
data['dob_years'].unique() 

среднее                15188
высшее                  5251
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

женат / замужем          12344
гражданский брак          4163
не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

children  days_employed  dob_years education  education_id  \
0             1          -8437         42    высшее             0   
1             1          -4024         36   среднее             1   
2             0          -5623         33   среднее             1   
3             3          -4124         32   среднее             1   
4             0         340266         53   среднее             1   
...         ...            ...        ...       ...           ...   
21449         1          -4529         43   среднее             1   
21450         0         343937         67   среднее             1   
21451         1          -2113         38   среднее             1   
21452         3          -3112         38   среднее             1   
21453         2          -1984         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21449  гражданский брак                 1      F   компаньон     0   
21450   женат / замужем                 0      F   пенсионер     0   
21451  гражданский брак                 1      M   сотрудник     1   
21452   женат / замужем                 0      M   сотрудник     1   
21453   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
0            253875                 покупка жилья  
1            112080       приобретение автомобиля  
2            145885                 покупка жилья  
3            267628    дополнительное образование  
4            158616               сыграть свадьбу  
...             ...                           ...  
21449        224791             операции с жильем  
21450        155999          сделка с автомобилем  
21451         89672                  недвижимость  
21452        244093  на покупку своего автомобиля  
21453         82047         на покупку автомобиля  

[21454 rows x 12 columns]

array([42, 36, 33, 32, 53, 27, 43, 50, 35, 41, 40, 65, 54, 56, 26, 48, 24,
       21, 57, 67, 28, 63, 62, 47, 34, 68, 25, 31, 30, 20, 49, 37, 45, 61,
       64, 44, 52, 46, 23, 38, 39, 51,  0, 59, 29, 60, 55, 58, 71, 22, 73,
       66, 69, 19, 72, 70, 74, 75])

In [17]:
print(data[data['education']=='среднее']['education_id'].count())
print(data[data['education']=='высшее']['education_id'].count())
print(data[data['education']=='неоконченное высшее']['education_id'].count())
print(data[data['education']=='начальное']['education_id'].count())
print(data[data['education']=='ученая степень']['education_id'].count())
data=data.drop(data[data['dob_years']==0].index)


15172
5250
744
282
6


In [18]:
display(data.head(10))
data['dob_years'].unique()

children  days_employed  dob_years education  education_id  \
0         1          -8437         42    высшее             0   
1         1          -4024         36   среднее             1   
2         0          -5623         33   среднее             1   
3         3          -4124         32   среднее             1   
4         0         340266         53   среднее             1   
5         0           -926         27    высшее             0   
6         0          -2879         43    высшее             0   
7         0           -152         50   среднее             1   
8         2          -6929         35    высшее             0   
9         0          -2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

array([42, 36, 33, 32, 53, 27, 43, 50, 35, 41, 40, 65, 54, 56, 26, 48, 24,
       21, 57, 67, 28, 63, 62, 47, 34, 68, 25, 31, 30, 20, 49, 37, 45, 61,
       64, 44, 52, 46, 23, 38, 39, 51, 59, 29, 60, 55, 58, 71, 22, 73, 66,
       69, 19, 72, 70, 74, 75])

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

In [19]:
new_education=data[['education','education_id']]
display(new_education)

education  education_id
0        высшее             0
1       среднее             1
2       среднее             1
3       среднее             1
4       среднее             1
...         ...           ...
21449   среднее             1
21450   среднее             1
21451   среднее             1
21452   среднее             1
21453   среднее             1

[21353 rows x 2 columns]

In [20]:
new_family = data[['family_status', 'family_status_id']]
display(new_family)
display(new_family.value_counts())

family_status  family_status_id
0       женат / замужем                 0
1       женат / замужем                 0
2       женат / замужем                 0
3       женат / замужем                 0
4      гражданский брак                 1
...                 ...               ...
21449  гражданский брак                 1
21450   женат / замужем                 0
21451  гражданский брак                 1
21452   женат / замужем                 0
21453   женат / замужем                 0

[21353 rows x 2 columns]

family_status          family_status_id
женат / замужем        0                   12290
гражданский брак       1                    4130
не женат / не замужем  4                    2794
в разводе              3                    1185
вдовец / вдова         2                     954
dtype: int64

In [21]:
data=data.drop(['education','family_status'], axis=1)
display(data.sort_values('children'))

children  days_employed  dob_years  education_id  family_status_id  \
7718          0          -1963         54             0                 0   
10056         0          -2634         42             0                 1   
10057         0          -5958         57             0                 0   
10058         0          -8642         57             1                 0   
10059         0          -1188         39             1                 1   
...         ...            ...        ...           ...               ...   
15787         5           -418         31             1                 0   
20770         5          -2386         35             1                 0   
7859          5           -773         36             1                 0   
3977          5         191683         42             1                 1   
16173         5           -387         35             1                 1   

      gender  income_type  debt  total_income  \
7718       F    сотрудник     0        136493   
10056      M  госслужащий     0        252707   
10057      F    сотрудник     0         77913   
10058      M    компаньон     0        205397   
10059      M    компаньон     0        201279   
...      ...          ...   ...           ...   
15787      F    сотрудник     0         77552   
20770      F    компаньон     0        204241   
7859       F    сотрудник     0         48772   
3977       M    сотрудник     0        145017   
16173      F  госслужащий     0        126102   

                                     purpose  
7718   получение дополнительного образования  
10056                        сыграть свадьбу  
10057                             автомобиль  
10058             дополнительное образование  
10059                                свадьба  
...                                      ...  
15787       сделка с подержанным автомобилем  
20770                                  жилье  
7859                       операции с жильем  
3977            на покупку своего автомобиля  
16173                  на проведение свадьбы  

[21353 rows x 10 columns]

### Шаг 2.6. Категоризация дохода.

In [22]:
def price(kosten):
    if kosten<=30000:
        return 'E'
    elif 3000<kosten<=50000:
        return 'D'
    elif 50000<kosten<=200000:
        return 'C'
    elif 200000<kosten<=1000000:
        return 'B'
    else:
        return 'A'
data['price']=data['total_income'].apply(lambda kosten:price(kosten))

def dob_years_group(age):
    if 19<=age<=35:
        return 'молодые'
    if 36<age<59:
        return 'взрослые'
    return 'пенсионеры'
data['dob_years_category']=data['dob_years'].apply(lambda age:dob_years_group(age))
display(data)

children  days_employed  dob_years  education_id  family_status_id  \
0             1          -8437         42             0                 0   
1             1          -4024         36             1                 0   
2             0          -5623         33             1                 0   
3             3          -4124         32             1                 0   
4             0         340266         53             1                 1   
...         ...            ...        ...           ...               ...   
21449         1          -4529         43             1                 1   
21450         0         343937         67             1                 0   
21451         1          -2113         38             1                 1   
21452         3          -3112         38             1                 0   
21453         2          -1984         40             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          F   сотрудник     0        253875                 покупка жилья   
1          F   сотрудник     0        112080       приобретение автомобиля   
2          M   сотрудник     0        145885                 покупка жилья   
3          M   сотрудник     0        267628    дополнительное образование   
4          F   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21449      F   компаньон     0        224791             операции с жильем   
21450      F   пенсионер     0        155999          сделка с автомобилем   
21451      M   сотрудник     1         89672                  недвижимость   
21452      M   сотрудник     1        244093  на покупку своего автомобиля   
21453      F   сотрудник     0         82047         на покупку автомобиля   

      price dob_years_category  
0         B           взрослые  
1         C         пенсионеры  
2         C            молодые  
3         B            молодые  
4         C           взрослые  
...     ...                ...  
21449     B           взрослые  
21450     C         пенсионеры  
21451     C           взрослые  
21452     B           взрослые  
21453     C           взрослые  

[21353 rows x 12 columns]

В условии к задаче не было указано о создании стобца dob_years_category
но когда дошел до обработки данных эта групировка очень помогла.

### Шаг 2.7. Категоризация целей кредита.

In [23]:
display(data['purpose'].value_counts())


свадьба                                   786
на проведение свадьбы                     764
сыграть свадьбу                           760
операции с недвижимостью                  672
покупка коммерческой недвижимости         658
покупка жилья для сдачи                   649
операции с коммерческой недвижимостью     648
операции с жильем                         646
покупка жилья                             640
жилье                                     640
покупка жилья для семьи                   637
строительство собственной недвижимости    633
недвижимость                              629
операции со своей недвижимостью           627
строительство жилой недвижимости          621
строительство недвижимости                619
покупка своего жилья                      619
покупка недвижимости                      618
ремонт жилью                              605
покупка жилой недвижимости                603
на покупку своего автомобиля              502
заняться высшим образованием      

In [24]:
def purpose_new(row):
    if 'свадьб' in row:
        return 'проведение свадьбы'
    elif 'недвижимост' in row or 'жиль' in row:
        return 'операции с недвижимостью'
    elif 'образован' in row:
        return 'получение образования'
    elif 'автомоб' in row:
        return 'операции с автомобилем'
data['purpose_category']=data['purpose'].apply(lambda row:purpose_new(row))

display(data)

children  days_employed  dob_years  education_id  family_status_id  \
0             1          -8437         42             0                 0   
1             1          -4024         36             1                 0   
2             0          -5623         33             1                 0   
3             3          -4124         32             1                 0   
4             0         340266         53             1                 1   
...         ...            ...        ...           ...               ...   
21449         1          -4529         43             1                 1   
21450         0         343937         67             1                 0   
21451         1          -2113         38             1                 1   
21452         3          -3112         38             1                 0   
21453         2          -1984         40             1                 0   

      gender income_type  debt  total_income                       purpose  \
0          F   сотрудник     0        253875                 покупка жилья   
1          F   сотрудник     0        112080       приобретение автомобиля   
2          M   сотрудник     0        145885                 покупка жилья   
3          M   сотрудник     0        267628    дополнительное образование   
4          F   пенсионер     0        158616               сыграть свадьбу   
...      ...         ...   ...           ...                           ...   
21449      F   компаньон     0        224791             операции с жильем   
21450      F   пенсионер     0        155999          сделка с автомобилем   
21451      M   сотрудник     1         89672                  недвижимость   
21452      M   сотрудник     1        244093  на покупку своего автомобиля   
21453      F   сотрудник     0         82047         на покупку автомобиля   

      price dob_years_category          purpose_category  
0         B           взрослые  операции с недвижимостью  
1         C         пенсионеры    операции с автомобилем  
2         C            молодые  операции с недвижимостью  
3         B            молодые     получение образования  
4         C           взрослые        проведение свадьбы  
...     ...                ...                       ...  
21449     B           взрослые  операции с недвижимостью  
21450     C         пенсионеры    операции с автомобилем  
21451     C           взрослые  операции с недвижимостью  
21452     B           взрослые    операции с автомобилем  
21453     C           взрослые    операции с автомобилем  

[21353 rows x 13 columns]

### Ответы на вопросы.

##### Вопрос 1:Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [25]:
chil_debt=data.groupby('children').agg({'debt':['count','sum']})
display(chil_debt)
chil_quality=chil_debt['debt']['sum']/chil_debt['debt']['count']
print(chil_quality*100)
def ch3(count):
    if count ==0:
        return 'A'
    elif count ==1:
        return 'B'
    elif count==2:
        return 'C'
    else:
        return 'D'
data['ch_category']=data['children'].apply(lambda count:ch3(count))
#display(data)
chil_debt1=data.groupby('ch_category').agg({'debt':['count','sum']})
display(chil_debt1)
chil_quality1=chil_debt1['debt']['sum']/chil_debt1['debt']['count']
display(chil_quality1*100)

debt        
          count     sum
children               
0         14022  1058.0
1          4839   442.0
2          2114   202.0
3           328    27.0
4            41     4.0
5             9     0.0

children
0    7.545286
1    9.134119
2    9.555345
3    8.231707
4    9.756098
5    0.000000
dtype: float64


debt        
             count     sum
ch_category               
A            14022  1058.0
B             4839   442.0
C             2114   202.0
D              378    31.0

ch_category
A    7.545286
B    9.134119
C    9.555345
D    8.201058
dtype: float64

По данным получается, что люди у которых  нет детей меньше всех просрачивают платеж, когда люди у которых 1,2 и 4 чаще остальных опоздывают с платежом по кредиту, и чуть меньше опоздывают люди у которых 3 ребенка. Вывод на глазах, дети влияют на своевременный возврат кредита, и чем больше их тем чаще нарушается график хотя и не существенно.

По попводу 5х детей не до конца понял почему тут 0 ноль появился, хотя проверял значения, может чего пропустил?

Есть ли зависимость между семейным положением и возвратом кредита в срок?


In [26]:
family_grouped=data.groupby('family_status_id').agg({'debt':['count','sum']}) 
family_debt_quality=family_grouped['debt']['sum']/family_grouped['debt']['count']
display(family_grouped)
display(new_family.value_counts())
display(family_debt_quality*100)

debt       
                  count    sum
family_status_id              
0                 12290  927.0
1                  4130  386.0
2                   954   62.0
3                  1185   85.0
4                  2794  273.0

family_status          family_status_id
женат / замужем        0                   12290
гражданский брак       1                    4130
не женат / не замужем  4                    2794
в разводе              3                    1185
вдовец / вдова         2                     954
dtype: int64

family_status_id
0    7.542718
1    9.346247
2    6.498952
3    7.172996
4    9.770938
dtype: float64

Из полученных данных выходи, что чаще всех просрачивают платеж по кредиту люди состоящие в гражданском браке 
и не женаты/не замужные, что собственно не удивительно)
Реже опаздывают с платежом женаты/замужные и те кто в разводе, и реже всех просрачивают платеж вдовцы и вдовы.
На вопрос есть ли связь между семейным положением и возвратом кредита - она есть, чем меньше у людей ответственности
а это не замужные/не женатые и состоящие в гражданском браке тем чаще опаздывают с платежом.


Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [27]:
total_income_grouped=data.groupby('price').agg({'debt':['count','sum']}) 
total_debt_quality=total_income_grouped['debt']['sum']/total_income_grouped['debt']['count']
display(total_income_grouped['debt'])
display(total_debt_quality*100)

count     sum
price               
A         25     2.0
B       5015   355.0
C      15943  1353.0
D        348    21.0
E         22     2.0

price
A    8.000000
B    7.078764
C    8.486483
D    6.034483
E    9.090909
dtype: float64

Значений для низкого и высокго заработка не большие и не думаю что они чтото покажут, а вот люди выше среднего до 1000000 - 'B'
где то в середине списка лишь 7% , люди с заработком до 200000 -'C'- чаще просрачивают платеж,и и еще реже люди с заработком до
50000 -'D'. 
Етсь ли зависимость между заработком и вовремя вернуть кредит, да и  зависит от уровня достатка, люди с заработком выше 
среднего и ниже меньше просрачивают по платежам, хотя люди со среднем достатком чаще грешат забывчивостью.

Как разные цели кредита влияют на его возврат в срок?

In [28]:
data_pivot=data.pivot_table(index=['dob_years_category'],columns='price',values='debt',aggfunc='sum')
display(data_pivot)

price                 A      B      C    D    E
dob_years_category                             
взрослые            2.0  189.0  638.0  9.0  1.0
молодые             0.0  133.0  560.0  5.0  0.0
пенсионеры          0.0   33.0  155.0  7.0  1.0

Из сводной таблицы выходит что чаще всего опаздывают с платежом взрослые от 39 лет люди среднего достатка до 200000, примерно так же 
забывают оплатить кредит молодые люди срденего достатка, и следом но реже люди с заработком выше среднего от 39 лет до 1000000.
В основном просрачивает средний сегмент что наверно обусловлено экономией средств, тогда как люди обеспеченые почти не забывают.

In [29]:
data_pivot=data.pivot_table(index=['income_type','dob_years_category'],columns='price',values='debt',aggfunc='sum')
display(data_pivot)

price                                 A      B      C    D    E
income_type     dob_years_category                             
безработный     взрослые            NaN    0.0    NaN  NaN  NaN
                молодые             NaN    NaN    1.0  NaN  NaN
в декрете       взрослые            NaN    NaN    1.0  NaN  NaN
госслужащий     взрослые            NaN   10.0   34.0  0.0  0.0
                молодые             NaN    8.0   28.0  1.0  NaN
                пенсионеры          NaN    0.0    5.0  NaN  NaN
компаньон       взрослые            2.0   60.0  113.0  1.0  0.0
                молодые             0.0   41.0  132.0  0.0  NaN
                пенсионеры          0.0    9.0   17.0  0.0  NaN
пенсионер       взрослые            NaN   18.0   79.0  2.0  1.0
                молодые             NaN    0.0    1.0  NaN  NaN
                пенсионеры          NaN   13.0   95.0  5.0  1.0
предприниматель взрослые            NaN    NaN    0.0  NaN  NaN
                молодые             NaN    0.0    NaN  NaN  NaN
сотрудник       взрослые            0.0  101.0  411.0  6.0  0.0
                молодые             0.0   84.0  398.0  4.0  0.0
                пенсионеры          0.0   11.0   38.0  2.0  NaN
студент         молодые             NaN    NaN    0.0  NaN  NaN

Место работы или ее отсутствие

In [30]:
purpose_grouped=data.groupby('purpose_category').agg({'debt':['sum','count']})
purpose_convers=(purpose_grouped['debt']['sum']/purpose_grouped['debt']['count'])*100
display(purpose_convers.sort_values())

purpose_category
операции с недвижимостью    7.237087
проведение свадьбы          7.965368
получение образования       9.261577
операции с автомобилем      9.337068
dtype: float64

цели кредита влияют на срок его погашения, и чаще других это кредиты на образование и автомобиль 9%,за проведение свадьбы и операции с недвижимостью кредиты чаще помнят и опаздывают лишь 7%

## Общий вывод:

    Проделанный анализ показывает, что определенный критерии и факторы такие как должность, возраст, семейное положение, количество детей или их отсутсвие влияет на срок возврата кредита.
    Если у человека нет детей то кредит возвращается реже, лишь в 7,5% случаях, тогда как люди с 1 ребенком и 2-мя чаще других опаздывают со сроком, 9,1% и 9,4% соответственно, и люди с 3 и более детьми просрачивают платеж в 8%, что подталкивает к выводу что люди 1 или 2 мя детьми чаще остальных забывают про кредитным обязательства.
    Возраст заемщика, место работы. и его финансовое положение не менее показателеен при возврате кредитных средств - чаще остальных опаздывают с возвратом денег это сотрудники в взрослом и молодом возрасте со средним достатком до 200000, далее следует категория команьоны с этих же возрастных категориях со средним достатком до 200000 и меньше остальных опаздывают с платежом пенсионеры и гослужащие во взрослом возрасте со средним и высоким достатком.
    Цели кредита так же имеют весомое значение при возврате вовремя кредитных средств. В случаях когда цель была операция с недвижимостью или на проведение свадьбы то заемщики не возвращали средства вовремя лишь в 7-8 % случаях, тогда как на образование или покупку автомобиля приводили к задержке до 9,3% случаях.